# Optional PyTorch Training Loop

Run a small neural training loop and save a checkpoint when PyTorch is installed. Without PyTorch, this cell prints install guidance and exits cleanly.

In [1]:
from pathlib import Path

exec(Path("samples/ml_tasks/torch_training_loop.py").read_text(encoding="utf-8"))


epoch   1: loss=0.7331 test_accuracy=0.4867
epoch  10: loss=0.6397 test_accuracy=0.7067
epoch  20: loss=0.5673 test_accuracy=0.7100
epoch  30: loss=0.5020 test_accuracy=0.7333
epoch  40: loss=0.4623 test_accuracy=0.7333
epoch  50: loss=0.4538 test_accuracy=0.7367
epoch  60: loss=0.4477 test_accuracy=0.7433
epoch  70: loss=0.4426 test_accuracy=0.7433
epoch  80: loss=0.4396 test_accuracy=0.7500
checkpoint: artifacts/ml_tasks/torch_model.pt


## Finding: CLI run

- `run-clean` completed with status `ok` in 6.771 seconds and executed 1 of 1 code cells.
- The PyTorch loop produced usable progress output from epoch 1 through epoch 80; final logged metrics were `loss=0.4396` and `test_accuracy=0.7500`.
- Checkpoint artifact produced: `artifacts/ml_tasks/torch_model.pt`.
- `show-cell --outputs full` made the longer multiline training log easy to inspect after execution. `run-clean` did not stream intermediate progress to the terminal during execution, so long or unstable training jobs would benefit from live progress visibility or a tailing mode.
